File: backend/src/main/java/com/example/backend/service/MapService.java
package com.example.backend.service;

import java.util.List;

import com.example.backend.endpoint.dto.CityDto;
import com.example.backend.endpoint.dto.CreateMapDto;
import com.example.backend.endpoint.dto.MapPointDto;
import com.example.backend.endpoint.dto.PDFDto;
import com.example.backend.endpoint.dto.StatusDto;

public interface MapService {


  /**
   * Creates a new map in the system.
   *
   * @param createMapDto dto which contains all needed properties
   * @return an CreateMapDto of the saved map
   */
  CreateMapDto create(CreateMapDto createMapDto);

  /**
   * Get all maps saved in the system
   *
   * @return a list of CreateMapDtos of all saved maps
   */
  List<CreateMapDto> getAllMaps();

  /**
   * Gets the cities of the Map from OpenStreetMap/Overpass API
   *
   * @param id of the Map
   * @return a list of Cities
   */
  List<CityDto> getInitialCities(Long id);

  /**
   * Gets the cities of the Map from local storage
   *
   * @param id of the Map
   * @return a list of Cities
   */
  List<CityDto> getCities(Long id);


  /**
   * Gets the towns of the Map from OpenStreetMap/Overpass API
   *
   * @param id of the Map
   * @return a list of Cities
   */
  List<CityDto> getTowns(Long id);

  /**
   * Creates the MapPoints for further processing of a map with the chosen cities
   *
   * @param id id of the corresponding Map
   * @param cityDtos list of the chosen cities
   * @param save if the MapPoints should be saved, or they are just for viewing purposes (when still selecting cities)
   * @return the MapPoints used for visualizing the connections between cities
   */
  List<MapPointDto> createMapPoints(Long id, List<CityDto> cityDtos, boolean save);

  /**
   * Colorizes the MapPoints of a map with the chosen cities and saves them to the map
   *
   * @param id id of the corresponding Map
   * @param cityDtos list of the chosen cities
   * @return the MapPoints used for visualizing the colored connections between cities
   */
  List<MapPointDto> colorizeMapPoints(Long id, List<CityDto> cityDtos);


  /**
   * Gets the Map from local storage
   *
   * @param id of the Map
   * @return a CreateMapDto
   */
    CreateMapDto get(Long id);

  /**
   * Gets all MapPoints from a certain Map
   *
   * @param id of the Map
   * @return a List of MapPointDtos
   */
    List<MapPointDto> getMapPoints(Long id);

  /**
   * Creates the GameBoard and TicketCards for a saved Map and its corresponding MapPoints
   *
   * @param id of the Map
   * @param DPI print resolution of OSM image
   * @return a PDFDto containing the GameBoard and its corresponding TicketCards
   */
  PDFDto createGameBoard(Long id, int DPI);

   /**
   * Gets the GameBoard and Tickets as a PDFDto from as saved Map
   *
   * @param id of the Map
   * @return a PDFDto containing the GameBoard and its corresponding TicketCards
   */
  PDFDto getGameBoard(Long id);


  /**
   * Gets the status of a Maps progress when creating tickets and gameboards
   *
   * @param id of the Map
   * @return a StatusDto that contains a progress message
   */
  StatusDto getStatus(Long id);
}

File: frontend/src/app/services/map.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {MapDto} from '../dtos/map';
import {Observable} from "rxjs";
import {City} from "../dtos/city";
import {MapPointDto} from "../dtos/map-point";
import {PDFDto} from "../dtos/pdf";
import {StatusDto} from "../dtos/status";

@Injectable({
  providedIn: 'root'
})
export class MapService {

  private baseURI = 'http://127.0.0.1:8080/api';

  constructor(private httpClient: HttpClient) {
  }


  /**
   * creates a map in the system
   *
   * @param map map to create in the system
   *
   * @returns an mapDto representation of the registered map
   */
  createMap(map: MapDto): Observable<MapDto> {
    return this.httpClient.post<MapDto>(this.baseURI + '/maps', map);
  }

  /**
   * loads all maps from the system
   *
   * @returns a mapDto List of the registered maps
   */
  getAllMaps(): Observable<MapDto[]> {
    return this.httpClient.get<MapDto[]>(this.baseURI + '/maps');
  }

  /**
   * loads a map from the system
   *
   * @param id of map to fetch from the system
   *
   * @returns an mapDto representation of the registered map
   */
  getMap(id: number): Observable<MapDto> {
    return this.httpClient.get<MapDto>(this.baseURI + '/maps/' + id);
  }

  /**
   * gets all cities belonging to map
   *
   * @param id of map saved on server
   *
   * @returns a list of cites
   */
  getCities(id: number): Observable<City[]> {
    return this.httpClient.get<City[]>(this.baseURI + '/maps/cities/' + id);
  }

  /**
   * gets all towns belonging to map
   *
   * @param id of map saved on server
   *
   * @returns a list of cities (towns are also saved as cities)
   */
  getTowns(id: number): Observable<City[]> {
    return this.httpClient.get<City[]>(this.baseURI + '/maps/towns/' + id);
  }

  /**
   * shows current chosen cities as MapPoints. MapPoints are used for later calculations
   *
   * @param id of map saved on server
   * @param cities list of chosen cities
   *
   * @returns a list of mapPoints
   */
  showMapPoints(id: number, cities: City[]): Observable<MapPointDto[]> {
    return this.httpClient.post<MapPointDto[]>(this.baseURI + '/maps/selection/' + id, cities);
  }

  /**
   * colorizes the prior chosen MapPoints and saves them to the map
   *
   * @param id of map saved on server
   * @param cities list of chosen cities
   *
   * @returns a list of mapPoints
   */
  colorizeMapPoints(id: number, cities: City[]): Observable<MapPointDto[]> {
    return this.httpClient.post<MapPointDto[]>(this.baseURI + '/maps/colorization/' + id, cities);
  }

  /**
   * gets all MapPoints of a certain map
   *
   * @param id of map saved on server
   *
   * @returns a list of mapPoints corresponding to the map
   */
  getMapPoints(id: number): Observable<MapPointDto[]> {
    return this.httpClient.get<MapPointDto[]>(this.baseURI + '/maps/mapPoints/' + id);
  }

  /**
   *
   * creates the gameBoard and its ticketCards of a certain map
   *
   * @param id of map saved on server
   * @param DPI render resolution for OSM Map-Image
   *
   * @returns a PDFDto containing gameBoard and ticketCards of the map
   */
  createGameBoard(id: number, map: MapDto): Observable<PDFDto> {
    return this.httpClient.post<PDFDto>(this.baseURI + '/maps/gameBoard/create/'+id,map);
  }

  /**
   *
   * gets the gameBoard and its ticketCards of a certain map
   *
   * @param id of map saved on server
   *
   * @returns a PDFDto containing gameBoard and ticketCards of the map
   */
  getGameBoard(id: number): Observable<PDFDto> {
    return this.httpClient.get<PDFDto>(this.baseURI + '/maps/gameBoard/'+id);
  }


  /**
   * Gets the status of a Maps progress when creating tickets and gameboards
   *
   * @param id of the Map
   * @return a StatusDto that contains a progress message
   */
  getStatus(id:number): Observable<StatusDto> {
    return this.httpClient.get<StatusDto>(this.baseURI + '/maps/status/' +id);
  }
}
